commons-annotations/src/main/scala/com/avsystem/commons/rpc/rpcAnnotations.scala

```diff
@@ -0,0 +1,422 @@
+  *   private def readJson[T: GenCodec](json: Json): T =
+  *     JsonStringInput.read[T](json.jsonStr)
+  *
+  *   private def writeJson[T: GenCodec](value: T): Json =
+  *     Json(JsonStringOutput.write[T](value))
+  *
+  *   implicit def genCodecBasedJsonEncoding[T: GenCodec]: AsRawReal[T,Json] =
+  *     AsRawReal.create[T,Json](readJson[T], writeJson[T])
+  *
+  *   // instead of using `mapNow`, this method can also take implicit ExecutionContext and just use `map`
+  *   implicit def genCodecBasedFutureJsonEncoding[T: GenCodec]: AsRawReal[Future[Json],Future[T]] =
+  *     AsRawReal.create[Future[T],Future[Json]](_.mapNow(readJson[T]), _.mapNow(writeJson[T]))
+  * }
+  *
+  * trait AsyncRawRpc {
+  *   def call(rpcName: String, @multi args: Map[String,Json]): Future[Json]
+  * }
+  * }}}
+  *
+  * If you don't want to introduce the wrapper `Json` class and use more raw type, e.g. plain `String` then you
+  * can also do it by moving implicit instances of `AsReal` and `AsRaw` (or the joined `AsRawReal`) into the
+  * `implicits` object of raw RPC companion:
+  *
+  * {{{
+  * trait AsyncRawRpc {
+  *   def call(rpcName: String, @multi args: Map[String,String]): Future[String]
+  * }
+  * object AsyncRawRpc extends RawRpcCompanion[AsyncRawRpc] {
+  *   private def readJson[T: GenCodec](json: String): T =
+  *     JsonStringInput.read[T](json)
+  *   private def writeJson[T: GenCodec](value: T): String =
+  *     JsonStringOutput.write[T](value)
+  *
+  *   override object implicits {
+  *     implicit def genCodecBasedJsonEncoding[T: GenCodec]: AsRawReal[String,T] =
+  *       AsRawReal.create[String,T](readJson[T], writeJson[T])
+  *     implicit def genCodecBasedFutureJsonEncoding[T: GenCodec]: AsRawReal[Future[String],Future[T]] =
+  *       AsRawReal.create[Future[String],Future[T]](_.mapNow(readJson[T]), _.mapNow(writeJson[T]))
+  *   }
+  * }
+  * }}}
+  */
+final class encoded extends RpcEncoding
+
+/**
+  * Turns off raw value encoding as specified by [[encoded]]. By default, [[single]] and [[optional]] raw parameters
+  * are already [[verbatim]], so using [[verbatim]] only makes sense on [[multi]] raw parameters or
+  * raw methods themselves, which means turning off encoding of method's result.
+  *
+  * When encoding is turned off, raw and real types must be exactly the same types. For example, the following raw RPC
+  * definition will match only raw RPC traits whose methods take `Int`s as parameters and return `Double`s as values:
+  *
+  * {{{
+  * trait VerbatimRawRpc {
+  *   @verbatim def call(rpcName: String, @multi @verbatim args: Map[String,Int]): Double
+  * }
+  * }}}
+  */
+final class verbatim extends RpcEncoding
+
+/**
+  * Method tagging lets you have more explicit control over which raw methods can match which real methods.
+  * Example:
+  *
+  * {{{
+  * sealed trait RestMethod extends RpcTag
+  * class GET extends RestMethod
+  * class POST extends RestMethod
+  *
+  * @methodTag[RestMethod,GET]
```

Comment:
Maybe default tag is not so useful after all? It seems like a good practice to explicitly write tags anyway (see below), and it's ugly when you don't want to introduce one.

Reply:
If you want all tags to be explicit, you can do it by providing base tag type as the default tag type, which means "no default tag":

```scala
@methodTag[RestMethod,RestMethod]
```

And then every raw method should have `@tagged[Something]` so that every real method must have explicit tag annotation.